diff --color -U 5 -r -Z -B ./org/gtk/gtkwindow.c ./mod/gtk/gtkwindow.c
--- ./org/gtk/gtkwindow.c	2019-09-03 21:53:07.000000000 +0300
+++ ./mod/gtk/gtkwindow.c	2019-12-29 16:24:09.937549137 +0200
@@ -4160,10 +4160,22 @@
     {
       gtk_style_context_add_class (gtk_widget_get_style_context (widget), "solid-csd");
     }
 
   priv->client_decorated = TRUE;
+
+  gboolean is_gnome_wayland = g_strcmp0 (g_getenv ("DESKTOP_SESSION"), "gnome") == 0;
+#ifdef GDK_WINDOWING_X11
+  is_gnome_wayland = is_gnome_wayland && !GDK_IS_X11_DISPLAY (gtk_widget_get_display (widget));
+#endif
+
+  if (!is_gnome_wayland && g_getenv("GTK_CSD") == FALSE)
+    {
+      gtk_style_context_remove_class (gtk_widget_get_style_context (widget), GTK_STYLE_CLASS_CSD);
+      gtk_style_context_remove_class (gtk_widget_get_style_context (widget), "solid-csd");
+      priv->client_decorated = FALSE;
+    }
 }
 
 static void
 on_titlebar_title_notify (GtkHeaderBar *titlebar,
                           GParamSpec   *pspec,
Only in ./mod/gtk: gtkwindow.c.orig
